pallets/dkg/src/functions.rs

```diff
@@ -0,0 +1,242 @@
+	/// an error.
+	pub fn verify(data: JobResult) -> DispatchResult {
+		match data {
+			JobResult::DKG(info) => Self::verify_generated_dkg_key(info),
+			JobResult::DKGSignature(info) => Self::verify_dkg_signature(info),
+			_ => Err(Error::<T>::InvalidJobType.into()), // this should never happen
+		}
+	}
+
+	/// Verifies the generated DKG key based on the provided DKG verification information.
+	///
+	/// The verification process includes generating required signers, validating signatures, and
+	/// ensuring a sufficient number of unique signers are present.
+	///
+	/// # Arguments
+	///
+	/// * `data` - The DKG verification information containing participants, keys, and signatures.
+	///
+	/// # Returns
+	///
+	/// Returns a `DispatchResult` indicating whether the DKG key verification was successful or
+	/// encountered an error.
+	fn verify_generated_dkg_key(data: DKGResult) -> DispatchResult {
+		// generate the required signers
+		let maybe_signers = data
+			.participants
+			.iter()
+			.map(|x| {
+				ecdsa::Public(
+					Self::to_slice_33(&x.encode())
+						.unwrap_or_else(|| panic!("Failed to convert input to ecdsa public key")),
+				)
+			})
+			.collect::<Vec<ecdsa::Public>>();
+
+		ensure!(!maybe_signers.is_empty(), Error::<T>::NoParticipantsFound);
+		ensure!(!data.keys_and_signatures.is_empty(), Error::<T>::NoSignaturesFound);
+
+		let mut known_signers: Vec<ecdsa::Public> = Default::default();
+		let signed_pub_key: Vec<u8> =
+			data.keys_and_signatures.first().expect("Cannot be empty").clone().0;
+
+		for (key, signature) in data.keys_and_signatures {
+			// ensure the required signer signature exists
+			let (maybe_authority, success) =
+				Self::verify_signer_from_set_ecdsa(maybe_signers.clone(), &key, &signature);
+
+			if success {
+				// sanity check, everyone signed the same key
+				ensure!(key == signed_pub_key, Error::<T>::InvalidSignatureData);
+
+				let authority = maybe_authority.ok_or(Error::<T>::CannotRetreiveSigner)?;
+
+				ensure!(!known_signers.contains(&authority), Error::<T>::DuplicateSignature);
+
+				known_signers.push(authority);
+			}
+		}
+
+		ensure!(known_signers.len() > data.threshold.into(), Error::<T>::NotEnoughSigners);
+
+		Ok(())
+	}
+
+	/// Verifies the DKG signature result by recovering the ECDSA public key from the provided data
+	/// and signature.
+	///
+	/// This function checks whether the recovered public key matches the expected signing key,
+	/// ensuring the validity of the signature.
+	///
+	/// # Arguments
+	///
+	/// * `data` - The DKG signature result containing the message data and ECDSA signature.
+	///
+	/// # Returns
+	///
+	/// Returns a `DispatchResult` indicating whether the DKG signature verification was successful
+	/// or encountered an error.
+	fn verify_dkg_signature(data: DKGSignatureResult) -> DispatchResult {
+		let recovered_key = Self::recover_ecdsa_pub_key(&data.data, &data.signature)
```

Comment:
We will need to add key/signature type info to the SignatureResult object so we can match on `ECDSA` or `Schnorr`